dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -0,0 +1,141 @@
+﻿#include "pch.h"
+#include "PushNotificationManager.h"
+#include "PushNotificationManager.g.cpp"
+#include "PushNotificationCreateChannelResult.h"
+#include <winrt\windows.networking.pushnotifications.h>
+#include <winerror.h>
+#include <algorithm>
+#include <vector>
+#include "PushNotificationChannel.h"
+#include <wil/resource.h>
+
+using namespace winrt::Windows::Networking::PushNotifications;
+using namespace winrt;
+
+namespace winrt::Microsoft::ProjectReunion::implementation
+{
+    inline constexpr std::uint32_t c_maxBackoffSeconds{ 960 };
+
+    const HRESULT WNP_E_NOT_CONNECTED = MAKE_HRESULT(SEVERITY_ERROR, FACILITY_ITF, 0x880403E8L);
+    const HRESULT WNP_E_RECONNECTING = MAKE_HRESULT(SEVERITY_ERROR, FACILITY_ITF, 0x880403E9L);
+    const HRESULT WNP_E_BIND_USER_BUSY = MAKE_HRESULT(SEVERITY_ERROR, FACILITY_ITF, 0x880403FEL);
+
+    bool PushNotificationManager::isChannelRequestRetryable(const hresult& hr)
+    {
+        switch (hr)
+        {
+        case HRESULT_FROM_WIN32(ERROR_TIMEOUT):
+        case WNP_E_NOT_CONNECTED:
+        case WPN_E_OUTSTANDING_CHANNEL_REQUEST:
+        case WNP_E_RECONNECTING:
+        case WNP_E_BIND_USER_BUSY:
+        case RPC_S_SERVER_UNAVAILABLE:
+            return true;
+        default:
+            return false;
+        }
+    }
+
+    inline bool IsPackagedProcess()
+    {
+        uint32_t packageFullNameLength = 0;
+        const auto rc{ GetCurrentPackageFullName(&packageFullNameLength, nullptr) };
+
+        if (rc != ERROR_INSUFFICIENT_BUFFER)
+        {
+            winrt::throw_hresult(rc);
+        }
+        return rc == ERROR_INSUFFICIENT_BUFFER;
+    }
+
+    Windows::Foundation::IAsyncOperationWithProgress<Microsoft::ProjectReunion::PushNotificationCreateChannelResult, Microsoft::ProjectReunion::PushNotificationCreateChannelStatus> PushNotificationManager::CreateChannelAsync(winrt::guid remoteId)
+    {
+        static std::vector<winrt::guid> s_remoteIdList;
+        static wil::critical_section s_lock;
+
+        // NOTE: API supports channel requests only for packaged applications
+        IsPackagedProcess();
+
+        check_pointer(&remoteId);
```

Comment:
I would flip this to be the very first check.

Reply:
Also GUID is not a pointer? So not sure how this would work. I think you meant to check for GUID_NULL?

Reply:
Good unit test should catch all these special cases

Reply:
@hulumane if it is not a packaged application why would we want to verify the validity of remoteId. In fact we should not be concerned about it right.

Reply:
Arguments always need to be validated in the beginning of the function or API call. 

RemoteId will be used for all Cobalt+ builds by calling into remote internal APIs irrespective of packaged or not. If the API does not exist in the SKU, we use the logic that we have below.